crates/bevy_pbr/src/extended_material.rs

```diff
@@ -0,0 +1,201 @@
+        SpecializedMeshPipelineError, UnpreparedBindGroup,
+    },
+    renderer::RenderDevice,
+    texture::{FallbackImage, Image},
+};
+
+use crate::{Material, MaterialPipeline, MaterialPipelineKey, StandardMaterial};
+
+/// A subset of the `Material` trait for use when extending `StandardMaterial`s.
+pub trait StandardMaterialExtension: Asset + AsBindGroup + Clone + Sized {
+    /// Returns this material's vertex shader. If [`ShaderRef::Default`] is returned, the default mesh vertex shader
+    /// will be used.
+    fn vertex_shader() -> ShaderRef {
+        ShaderRef::Default
+    }
+
+    /// Returns this material's fragment shader. If [`ShaderRef::Default`] is returned, the default mesh fragment shader
+    /// will be used.
+    #[allow(unused_variables)]
+    fn fragment_shader() -> ShaderRef {
+        ShaderRef::Default
+    }
+
+    /// Returns this material's prepass vertex shader. If [`ShaderRef::Default`] is returned, the default prepass vertex shader
+    /// will be used.
+    fn prepass_vertex_shader() -> ShaderRef {
+        ShaderRef::Default
+    }
+
+    /// Returns this material's prepass fragment shader. If [`ShaderRef::Default`] is returned, the default prepass fragment shader
+    /// will be used.
+    #[allow(unused_variables)]
+    fn prepass_fragment_shader() -> ShaderRef {
+        ShaderRef::Default
+    }
+
+    /// Customizes the default [`RenderPipelineDescriptor`] for a specific entity using the entity's
+    /// [`MaterialPipelineKey`] and [`MeshVertexBufferLayout`] as input.
+    /// Specialization for the `StandardMaterial` is applied before this function is called.
+    #[allow(unused_variables)]
+    #[inline]
+    fn specialize(
+        pipeline: &MaterialPipeline<ExtendedMaterial<Self>>,
+        descriptor: &mut RenderPipelineDescriptor,
+        layout: &MeshVertexBufferLayout,
+        key: MaterialPipelineKey<ExtendedMaterial<Self>>,
+    ) -> Result<(), SpecializedMeshPipelineError> {
+        Ok(())
+    }
+}
+
+/// A material that extends the [`StandardMaterial`] with user-defined shaders
+/// and data.
+/// The data from both the parts will be combined and made available to the shader
+/// so that the builtin pbr shader functions will work as expected, and custom
+/// data can also be used.
+/// If the material `M` returns a non-default result from `vertex_shader()` it will be used in place of the standard
+/// vertex shader (`bevy_pbr::render::mesh.wgsl`). It should return the same output as the standard vertex shader.
+/// If the material `M` returns a non-default result from `fragment_shader()` it will be used in place of the standard
+/// fragment shader (`bevy_pbr::render::pbr_fragment.wgsl`). since all the standard material fields are
+/// present, the `pbr_fragment` shader function can be called from the custom shader (see
+/// the `extended_material` example).
+#[derive(Asset, Clone, TypePath)]
+pub struct ExtendedMaterial<M: StandardMaterialExtension> {
```

Comment:
I feel like we should call it `ExtendedStandardMaterial`, as it only works with StandardMaterial (it would be nice if it was generic...).

Reply:
bit of a mouthful but you're right.

i thought generic wouldn't be possible due to orphan rules but now i can't see why. i'm going to have another quick go at that tonight.

Reply:
made it generic

Reply:
I agree and had the same thought, but felt `ExtendedMaterial` is maybe implicitly 'more' than `StandardMaterial`. You make a good point that as it only works on top of `StandardMaterial`, it would be better-named `ExtendedStandardMaterial`.

Reply:
it's no longer true - the extension can extend any material now, so i'm leaving this as is